### ISSUE TITLE
Keep track of Processor Vendor in AOT Validation

### ISSUE THREAD

**Issue by dsouzai** [eclipse/openj9#6836]
Opening this PR based on the following conversation in https://github.com/eclipse/omr/pull/4183

>andrewcraik commented 1 hour ago
>the AMD sequences are functionally different because we use multiple instructions rather than the single instruction on intel, but I'm fine to move that off to another place.
>
> 
>dsouzai commented 1 hour ago
>Does that mean if one was to generate (relocatable) code on Intel and run it on an AMD machine, the code wouldn't execute in an functionally equivalent way? Or is it just that the code sequence is no longer as optimal?
>
>andrewcraik commented 1 hour ago
>I believe it may not be functionally equivalent..

It may be functionally necessary to distinguish between processor vendors when validation AOT code. Because a SCC can be embedded in a container, it is possible for the SCC to have been generated on one processor and run on a different one.

**Comment by dsouzai** [comment]
@harryyu1994 does the new mechanism handle this automatically?

**Comment by harryyu1994** [comment]
no, we don't have checks for processor vendors. Currently we are only comparing the processor features for compatibility. This one definitely needs a work item.

**Comment by harryyu1994** [comment]
Does this mean the startup performance is going to suffer if the cluster has a mixture of AMD and Intel chips? This is not a good case for portability.

**Comment by dsouzai** [comment]
I guess this is a concern on all platforms; the compiler would need to be taught to only emit code that will run on all platforms of a given architecture. In the Intel vs AMD case, I believe the main difference is the NOP instructions that are emitted which are based on how they're expected to get patched. The rest are performance based decisions. I imagine this is something that could be resolved but it's probably best answered by @andrewcraik or @0xdaryl.

**Comment by harryyu1994** [comment]
This might be a big blow to portable AOT.

There could be two solutions to this:
1. The preferred solution is to generate portable AOT code that can run on both chips. 
2. If that's not possible, to get portable AOT, we might need to generate AOT code for both intel and AMD (i.e. double compile) while creating the SCC to be embedded in the container. This will double the AOT code storage in the SCC
which is probably not ideal.

**Comment by mpirvu** [comment]
I would strongly prefer (1), i.e. code that would work on both Intel and AMD given a set of features. This is important given the recent AMD resurgence. In Amazon AWS one could actually choose to run on AMD EPYC machines for 10% lower cost.

**Comment by mpirvu** [comment]
Scanning the code for `isAuthenticAMD()` I found 3 differences for AMD vs Intel:
1) `_paddingTable` is different. AMD is using `_K8PaddingTable` while Intel64 is using `_intelMultiBytePaddingTable`. Looking at the AMD manuals (see https://www.amd.com/system/files/TechDocs/47414_15h_sw_opt_guide.pdf chapter 5.8) they recommend the `_intelMultiBytePaddingTable` encoding for speed. I don't know if the usage of `_intelMultiBytePaddingTable` on AMD would would break our code patching scheme. If that's the case could we use `_K8PaddingTable` on Intel?
2) SSE instruction used for double precision. AMD prefers `MOVLPD` while Intel prefers `MOVSD`. The code seems to indicate that we could always use `MOVLPD` and be fine.
3) InstructionPatchAlignmentBoundary. Intel and AMD family15H use a size of 32 while everybody else uses 8. I think we need to update the code to include AMD family17H (ZEN2) and family 19H (ZEN3)

@andrewcraik @0xdaryl Do you think we can get away with common AOT code for Intel and AMD?

**Comment by harryyu1994** [comment]
Liberty-based Application (Acmeair)
1. Compile and fill up the Shared Class Cache with AOT on an Intel-based machine (500MB cache with 300MB AOT code), then move the Shared Class Cache to an AMD-based machine. Use `-Xjit:verbose` to make sure that we are actually loading AOT.
2. Compile on AMD-based machine and move to Intel-based machine.
- `-XX:+PortableSharedCache` worked in both cases.
